ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryBuilder.java

```diff
@@ -290,20 +290,33 @@ PersistentQueryMetadata buildPersistentQueryInDedicatedRuntime(
     final KeyFormat keyFormat;
     final ValueFormat valueFormat;
     final KsqlTopic ksqlTopic;
+    final KeyFormat pullQueryKeyFormat;
 
     switch (persistentQueryType) {
       // CREATE_SOURCE does not have a sink, so the schema is obtained from the query source
       case CREATE_SOURCE:
-        final DataSource dataSource = Iterables.getOnlyElement(sources);
+        DataSource dataSource = Iterables.getOnlyElement(sources);
 
         logicalSchema = dataSource.getSchema();
         keyFormat = dataSource.getKsqlTopic().getKeyFormat();
         valueFormat = dataSource.getKsqlTopic().getValueFormat();
         ksqlTopic = dataSource.getKsqlTopic();
-
+        pullQueryKeyFormat = dataSource.getKsqlTopic().getKeyFormat();
+        materializationProviderBuilderFactory.setPullQueryKeyFormat(pullQueryKeyFormat);
         break;
       default:
         logicalSchema = sinkDataSource.get().getSchema();
+        if (sinkDataSource.get().getKsqlTopic().getKeyFormat()
+                .getFormatInfo().getFormat().equalsIgnoreCase("JSON")
+                || sinkDataSource.get().getKsqlTopic().getValueFormat().getFormatInfo()
+                .getFormat().equalsIgnoreCase("JSON")
```

Comment:
why is the behavior different for topics that have JSON key / value? What does the value format have to do with it? Can we add tests that show this (e.g. avro key with JSON value)?

Reply:
For JSON format, it uses JsonSerde from sink and does not interact with schema registry. This keeps the existing logic.

Reply:
@hli21 but it wouldn't be a problem to treat JSON the same way, right? I think this special casing is pretty confusing.

Reply:
Remove this check.